### ISSUE TITLE
Why both "fg_inds.numel()=0 and bg_inds.numel()=0" not handled in proposal_target_layer.py?

### ISSUE THREAD

**Issue by hengck23** [ruotianluo/pytorch-faster-rcnn#11]
In proposal_target_layer.py,

```
  # Small modification to the original version where we ensure a fixed number of regions are sampled
  if fg_inds.numel() > 0 and bg_inds.numel() > 0:
    ...
  elif fg_inds.numel() > 0:
    ...
  elif bg_inds.numel() > 0:
    ...
  else:
    import pdb
    pdb.set_trace()

```

The case when both fg_inds.numel()=0 and bg_inds.numel()=0 is not implemented. 

When GT roi is not  used in training rcnn (set via the cfg TRAIN.USE_GT flag), fg_inds.numel() can be zero.
Note, this is not the case for original faster rcnn where  fg_inds.numel() != 0.

Also, the cfg TRAIN.BG_THRESH_HI  and TRAIN.BG_THRESH_LO values can also cause  bg_inds.numel()=0.

Since both "fg_inds.numel()=0 and bg_inds.numel()=0" is a possible case, I wonder why is it not handled?
In this case why don't just random select the roi and set them as negative?

**Comment by ruotianluo** [comment]
To be honest it's just copied from tf-faster-rcnn.

**Comment by jtdowdall** [comment]
I am experiencing this issue. I assume there is a negative sample in my dataset, how would I go about implementing a negative randomly selected roi?

**Comment by ruotianluo** [comment]
to_replace = all_rois.size(0) < rois_per_image
bg_inds = torch.from_numpy(npr.choice(np.arange(0, all_rois.size(0)), size=int(rois_per_image), replace=to_replace)).long().cuda()]
fg_rois_per_image = 0

try this, I didn't test; in principle it's extending bg_inds.numel() > 0.

**Comment by jtdowdall** [comment]
I had a negative value for a bounding box because of the 1-index correction in pascal_voc.py. Fixed that and it fixed this issue.

**Comment by mengce97** [comment]
i meet the same problem, how can i fix it?

**Comment by StrongBirds** [comment]
I have the same problem, how can I solve it

**Comment by RaphaelHighness** [comment]
Try to delete the folder named cache in this project and the folder which have the 'cache' in the name in VOCdevkit if applicable. I solve this problem but it can only train six iter and when the turn of  7th iter is coming the problem is back. (Pdb)... What's the problem and how to fix it? @ruotianluo

**Comment by mouyuan4598** [comment]
> ```
> to_replace = all_rois.size(0) < rois_per_image
> bg_inds = torch.from_numpy(npr.choice(np.arange(0, all_rois.size(0)), size=int(rois_per_image), replace=to_replace)).long().cuda()
> fg_rois_per_image = 0
> ```
> 
> try this, I didn't test; in principle it's extending bg_inds.numel() > 0.

@ruotianluo This additional condition is useful. It solves #68 as well. Please consider modify it in master code.

**Comment by ruotianluo** [comment]
Do you want to submit a pr?

**Comment by mouyuan4598** [comment]
> Do you want to submit a pr?

Ok, I will submit a pr.

**Comment by ruotianluo** [comment]
should be fixed by #169
